core/common/src/main/java/alluxio/uri/HostnameAuthority.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.uri;
+
+import alluxio.util.URIUtils;
+
+/**
+ * A hostname authority implementation.
+ */
+public class HostnameAuthority implements Authority {
+  private static final long serialVersionUID = 2580736424809131651L;
+
+  private final String mAuthority;
+
+  /**
+   * @param authority the authority string of the URI
+   */
+  public HostnameAuthority(String authority) {
+    mAuthority = authority;
+  }
+
+  @Override
+  public String getWholeAuthority() {
+    return mAuthority;
+  }
+
+  @Override
+  public String getConnectionAddress() {
+    return mAuthority;
+  }
+
+  @Override
+  public int compareTo(Authority other) {
```

Comment:
`mAuthority` and `other.mAuthority` may be null.
`compareTo` method does not support compare to null object.

Thanks

Reply:
It is a good practice to avoid using `null` as much as possible. When a field can be `null`, it requires special handling, and can potentially cause `NullPointerException` if we aren't careful. If we need to allow a field to be `null`, we annotate with `@Nullable` so that people know to be careful. Otherwise, the assumption is that it will not be `null`.

In this case, I don't see why we would ever want to create `HostnamePortAuthority` objects where the internal `String` value is `null`. I suggest preventing such objects from being created by using `Preconditions.checkNotNull(authority, "authority")` in the constructor. By checking once when the object is constructed, we no longer need to worry about the field being `null`.

Reply:
Now the `Authority`  will never be null. If it is null, then the `StandardURI.mAuthority` will be null directly. We will not create a `HostnamePortAuthority` or `ZookeeperAuthority` with `mAuthority == null`. This can help avoid many NPE and simplifies the logic.

Thanks!